convert2rhel/pkgmanager/handlers/dnf.py

```diff
@@ -0,0 +1,208 @@
+        self._base.conf.keepcache = True
+
+    def _enable_repos(self):
+        """Enable a list of required repositories."""
+        self._base.read_all_repos()
+        repos = self._base.repos.all()
+        enabled_repos = system_info.get_enabled_rhel_repos()
+        loggerinst.info("Enabling repositories:\n%s" % "\n".join(enabled_repos))
+        try:
+            for repo in repos:
+                # Disable the repositories that we don't want if the `repo.id`
+                # is not in the `enabled_repos` list, otherwise explicitly enable it
+                # to make sure that it will be available when we run the transactions.
+                repo.disable if repo.id not in enabled_repos else repo.enable()
+
+            # Load metadata of the enabled repositories
+            self._base.fill_sack()
+        except pkgmanager.exceptions.RepoError as e:
+            loggerinst.debug("Loading repository metadata failed: %s" % e)
+            loggerinst.critical("Failed to populate repository metadata.")
+
+    def _perform_operations(self):
+        """Perform the necessary operations in the transaction.
+
+        This internal method will actually perform three operations in the
+        transaction: update, reinstall and downgrade. The downgrade
+        will be executed only when the reinstall step raises
+        `PackagesNotAvailableError`.
+
+        :raises SystemExit: In case the dependency solving fails.
+        """
+        loggerinst.info("Performing operations on the transaction.")
+        original_os_pkgs = get_system_packages_for_replacement()
+        loggerinst.info("Performing upgrade, reinstall and downgrade of the %s packages ..." % system_info.name)
+        packages_to_remove = []
+        for pkg in original_os_pkgs:
+            self._base.upgrade(pkg_spec=pkg)
+            try:
+                self._base.reinstall(pkg_spec=pkg)
+            except pkgmanager.exceptions.PackagesNotAvailableError:
+                try:
+                    self._base.downgrade(pkg)
+                except pkgmanager.exceptions.PackagesNotInstalledError:
+                    loggerinst.warning("Package %s not available for downgrade.", pkg)
+                    packages_to_remove.append(pkg)
+
+        if packages_to_remove:
+            packages_to_remove = set(packages_to_remove)
+            loggerinst.debug(
+                "Removing problematic packages to continue with the conversion:\n%s", "\n".join(packages_to_remove)
+            )
+            remove_pkgs(
+                pkgs_to_remove=packages_to_remove,
+                backup=True,
+                critical=False,
+                reposdir=BACKUP_DIR,
+                varsdir=os.path.join(BACKUP_DIR, "dnf", "vars"),
+            )
+
+            loggerinst.debug("Finished backing up and removing the packages.")
+
+        loggerinst.debug("Finished update, reinstall, and downgrading of packages.")
```

Comment:
These messages about the operations are a bit misleading when we do the transaction validation only.
It'd be good to make them different when validating.

Reply:
I will take care of making them less confusing on Monday. Thanks for pointing that out.

Reply:
Note: This message needs to be updated to `yum` module as well.

Reply:
Looks good, thanks!
```
[10/12/2022 15:07:25] TASK - [Validate the dnf transaction]
Adding Oracle Linux Server packages to the dnf transaction set.
Resolving the dependencies of the packages in the dnf transaction set.
Downloading the packages that were added to the dnf transaction set.
Validating the dnf transaction set, no modifications to the system will happen this time.
Successfully validated the dnf transaction set.
[10/12/2022 15:14:39] DEBUG - Transaction validated successfully.

[10/12/2022 15:14:39] TASK - [Convert: Replace system packages] *********************************
Adding Oracle Linux Server packages to the dnf transaction set.
Resolving the dependencies of the packages in the dnf transaction set.
Downloading the packages that were added to the dnf transaction set.
Replacing Oracle Linux Server packages. This process may take some time to finish.
System packages replaced successfully.
```